### PR TITLE
Threads, jobs and safety maintenance.

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
Experiences from work on safe darktable closing and pthreads well behaving indicate to various bad dereferencing over the codebase if we are not in gui mode any more besides other more specific problems.

Here some code maintenance is done adding improved debugging logs also for release compilation, code readability and some dereferencing checks regarding an active control system and darktable gui.

@TurboGit quite some code changes, yet there is nothing tricky involved and is meant as a better base for testing issues discussed in #17984 . From the commit messages:
______________________________________________________________________________________
**Some jobs and control cleanup and fixes**

1. For code readability use 'control' instead of 's'
2. Simplify `dt_control_running()`
3. Two functions are strictly for internal use, so not exposed via .h file and becoming static
   - `_control_get_threadid()`
   - `_control_job_set_synchronous()`
4. Two function were not used at all (so were untested) so they were removed
   - `dt_control_job_set_state_callback()`
   - `dt_control_job_wait()`
5. Never create a control job if control is not in running state.
6. In control_jobs.c renamed all internal static functions by removing the leading 'dt' so they have a leading underscore for readability.

_________________________________________________________________
**pthreads related additions, maintenance and fixes**

1. As we use stdout for the logs we also use that here, make sure the logs are not corrupted by flushing.
2  Added `dt_pthread_join()` including error reports and assert() while in _DEBUG code. Make use of that in the codebase.
3. All `pthread_rwlock_t` related functions got variants for release builds with errors being reported.
4. As the penalty for using MUTEX_REPORTING is not measurable at all but gives additional safety and logs about runtime mutex related errors also in release build, we have this enabled at least until all pthread and control jobs related issues have been sorted out.
_________________________________________________________________________
**Some dereferencing safety**

1. While working on progress data we make sure the progress and control pointers are available.
2. If there is no imageio system available - was closed/removed before - we simply can't write any image file so do a quick exit without proceeding further.
3. Set darktable.lib to NULL after `free()`
4. Avoid dereferencing for `dt_view_manager_get_current_view()` and `dt_dev_gui_module()`
__________________________________________________________________________
**Introduce `dt_iop_has_focus()` and make use of it**

1. Implement a check for an iop module being in focus `dt_iop_has_focus()` and make use of that where possible for code deduplication
2. This check and requesting a module to be infocus includes a test for an available darktable.gui

### CODE REVIEW

**Comment by dterrahe** [comment]
> Some dereferencing safety

This is always a good goal, of course.

But it also makes sense to consider what guarantees are given about conditions persisting after initial tests and what the user would expect. For example, I personally would expect a command that was placed in the background to continue to completion unless I specifically cancelled it. Closing the application itself should not randomly pull out the rug from active background jobs, causing only partial completion and complicated error handling, and force them to check at each step if some framework they depend on has maybe been closed in the meantime.

> 2\. If there is no imageio system available - was closed/removed before

I would say `dt_imageio_cleanup` should only be called (in `dt_cleanup`) _after_ all background jobs have been stopped.
Similarly (probably) `dt_lib_cleanup`.

`gtk_widget_hide(dt_ui_main_window(darktable.gui->ui));` also hides the cancel buttons on background jobs, while it would be useful to still have access to those until they have all terminated.

In other words, work towards a desired program behavior rather than just plugging all the ways it can crash now because exit behavior is somewhat undefined.

**Comment by jenshannoschwalm** [comment]
@dterrahe good to see this being discussed :-)

I certainly don't want to pull all plugs simply to make it working somehow, for me it's more a *defensive* strategy - don't let it crash in an undefined state and leave notes in the log *why* and *where* bad things happen.

Also i think we agree that the mipmap, image_cache and imageio subsystems must be available while background / control jobs are still running.
dt UI in its current state is a mixed bag for me, we have all the responsiveness improvements delegating work from the gtk loop to background, unfortunately all the cleanup work has not been made fit for that. So dt now just crashes when closing via ctrl-q if there and running jobs using one of the above mentioned subsystems. (just for example we use the image_cache system and dereference on zero is not checked at many places ...)

  

> In other words, work towards a desired program behavior rather than just plugging all the ways it can crash now because exit behavior is somewhat undefined.

Absolutely.

> For example, I personally would expect a command that was placed in the background to continue to completion unless I specifically cancelled it.

I am not sure about this - although the above has been the behaviour all the time.
Just let us think about *printing, ex- importing* many images - it takes most time but certainly depending on the hardware drive connection, other background jobs as xmp writing can be pretty demanding too.

If i close **any** UI controlled application i personally would **not** expect it to be *running in the dark* for possibly a very long time. (How can i know a) when i can shutdown the machine, b) why dt can't be restarted because of the locks - was that due to a bug or because of pending work?)

So what is our **desired program behavior** after pressing crtl-q or click on close button? 
1. Keep the *UI open&active* until all pending background jobs have been done?
2. Leave a note on UI about a possibly time consuming work pending *in the dark* ?
3. Have some *pending work* info like partly implemented as 'closing splash'?
4. Understand the *closing action* as a user request to cancel all pending jobs, finish currently running jobs and close dt *asap*?

For me personally it would be either (4) or (1) ...

**Comment by HSUfineprint** [comment]
As an user, I personally prefer the software shows a small window with the message "darktable will close after XXX job is completed". If we want to close it without finishing pending jobs, there could be a button to do that.

**Comment by dterrahe** [comment]
1.) but with interaction limited to the "background jobs" panel (in order to allow cancellation), like `dt_gui_cursor_set_busy`. Add a toast in center to say that quitting is in progress waiting for active jobs. I don't think building a separate dialog (or reviving @ralfbrown 's closing splash for this purpose) is worth the effort of ironing out the kinks and unexpected complications. We really just want no surprises and no crashes. My expectation is that this is mostly relevant for tasks where people don't realise that for >items they may take a few seconds and are relegated to the background and immediately press quit but still logically assume their previous command was executed. If somebody _does_ want to spend extra time polishing the longer corner case, I'd say a dialog could pop up after 10 seconds saying there are still background jobs running and to cancel them in the jobs lib (or to offer a button to cancel all). But there are probably more important and lower hanging fruit in the ui that they could spend that time on.

> we use the image_cache system and dereference on zero is not checked at many places

Isn't it reasonable that these are guaranteed to be available before and as long as any image processing is done? None of that functions without the cache. How often are you going to check that the rug hasn't been pulled? Will every single access to an image pointer need to be mutexed? (I'm assuming you are not talking about memory allocation failures which "of course" should be checked).

**Comment by ralfbrown** [comment]
@jenshannoschwalm Of your four suggested possibilities, (3) is basically a variation of (2).  I obviously never finished the exit splash, but one thing I had in mind to do is have a button on it to hide the window.  So the user could opt to have the (much smaller) window stay on screen until darktable truly exited, or hide it right away after having been informed that work was being done in the dark.  (1) feels like terrible UX - "I wanted to close darktable, why isn't it closing?"  Expect lots of bug reports....  (4) is problematic because there are some background tasks we really shouldn't cancel on exit because they'll cause problems later, e.g. cancelling sidecar synch will cause the crawler to report changed sidecars on the next startup.

**Comment by dterrahe** [comment]
> So the user could opt to have the (much smaller) window stay on screen until darktable truly exited

Without a way to cancel background jobs? So that if the user pressed quit without realising they accidentally kicked off a long one, they now have no way to cancel it?

One person's terrible UX is another's handy feature. Would be nice to be able to fire off a long export and quit knowing that darktable will automatically close when done. Yes, you'd need a toast (at the minimum) to provide feedback. Did the current/old "I'll hide the window but continue running in the background so you can't restart, with a dense explanation about locks" lead to many bug reports?

**Comment by ralfbrown** [comment]
@dterrahe
The current situation is that there's no way to cancel a long-running task that the user kicked off just before typing Ctrl-Q, and no notice that there *is* a long-running task still in progress.  With (3), the user is informed that something will continue running and has to affirmatively dismiss the notice (on each exit) to hide the screen.  If we also reparent any progress bars to the exit screen, the user will even get an indication of how long the job(s) will take and the ability to cancel any cancellable jobs.  And if the user does *not* dismiss the exit screen, it will stay up until the remaining tasks are finished (the user might move other windows over top of it during that time, but it would still appear on the task bar and in the task switcher).

**Comment by jenshannoschwalm** [comment]
https://github.com/darktable-org/darktable/pull/18023/commits/08035391892b8b5d969eed0cb678d28d3278806e  implements a reordering of subsystems closing so all pending jobs keep going in the dark. The positive point - this could easily be done also for 5.01. The bad, for some reason the export job does not keep going so dt hangs. Would apreciate a hint to why/where-in-the-code ...

File: src/imageio/imageio.c [hunk 1]
```diff
@@ -1493,7 +1493,9 @@ gboolean dt_imageio_export_with_flags(const dt_imgid_t imgid,
   dt_dev_cleanup(&dev);
   dt_mipmap_cache_release(&buf);
 
-  if(!thumbnail_export && strcmp(format->mime(format_params), "memory")
+  if(dt_lua_running()
```
Comment (ralfbrown) [inline src/imageio/imageio.c:1496]: It looks like this will skip raising the signal if Lua is not compiled in, as well as not running.  Was that intended and correct behavior?

Comment (jenshannoschwalm) [reply src/imageio/imageio.c]: Indeed the signal raising can/should be done depending on dt_control_running() ... Fixed with latest force-push

**Comment by jenshannoschwalm** [comment]
I think the stability problems with dt crashing in pending backjobs are now finally solved and understood, the last two commits https://github.com/darktable-org/darktable/pull/18023/commits/fa19fddb2fdada44df9be80b92594311d6d4ff64 and https://github.com/darktable-org/darktable/pull/18023/commits/7c933ae1f60700c85e92b3ed5afe70739fdea540  are the most relevant ones. (There are some other issues with gui focus and image_cache not properly tested ...)

@TurboGit i think the whole bunch of commits is nothing for 5.01 as such a lot of code has been touched ... Let me know what commits might be ok for a release-fix are which aren't ...
 
There might also be some remaining lua issues not directly related with the shutdown problem and/or we might investigate later/elsewhere how to stop lua in a better way, @ralfbrown commented on this already and @wpferguson might have a look ...

This leaves out the **our desired program behavior** after pressing crtl-q part.

Technically i think we need/want for all 1-3 cases (it seems no one execpt me would opt for 4): 
a) a lock in dt_control_quit() so the relevant code get's only executed once to avoid race conditions
b) a pending-jobs counter
so i will do that first and add a big toast message about pending jobs in the back...

**Comment by wpferguson** [comment]
@ralfbrown the `dt_lua_running()` might be a better solution to the `_image_import_internal()` in common/image.c.  We can just add the test and not have to make other changes?

@jenshannoschwalm did you run some lua import scripts to test the changes?

**Comment by ralfbrown** [comment]
That's my conclusion as well - only try to take the lock if dt_lua_running returns TRUE.

**Comment by jenshannoschwalm** [comment]
Latest bunch of squash&force-pushed commits includes a) a protection against problems related to multiple ctrl-q signals and b) a working pending-jobs counter. Also a simple proposal for a delayed "closing window".

 @dterrahe and @ralfbrown what about this way? It's not exactly one of (1) to (4) but i think it might be good for us all ?

@wpferguson i did and couldn't spot problems yet  - but i use those scripts only occasionally and mostly while testing - so i can't confirm "everything is good" in a meaningful way.

Also i am not sure if the approach to test via `dt_lua_running()` before doing critical code is the `best way`. I mean, would it be possible to close the lua subsystem **after** `dt_control_shutdown()` as we do for imageio and the rest of the gang?

Anyway, can't trigger crashes or mutex/pthreads log errors any more ... so ready for intensive testing, review and suggestions.

**Comment by jenshannoschwalm** [comment]
@zisoft you might also want to test a _DEBUG build on your system, i tried that here also i couldn't find issues ...

**Comment by ralfbrown** [comment]
I don't think your proposal from 4fc9235 will come across at all well with average users:
1. they click the close button or type Ctrl-Q and nothing obvious happens (because they didn't notice the toast), then retry again, and again, getting more and more frustrated "why isn't it closing!"
2. they accidentally type Ctrl-Q (I've done that a number of times when meaning to type Ctrl-W), don't notice the toast, and 30 seconds later the program just suddenly disappears on them.  Cue crash report...

Something obvious and unmissable needs to happen immediately on a close request.

**Comment by zisoft** [comment]
> @zisoft you might also want to test a _DEBUG build on your system, i tried that here also i couldn't find issues ...

Will test intensively tomorrow and report back

**Comment by wpferguson** [comment]
> Also i am not sure if the approach to test via dt_lua_running() before doing critical code is the best way. I mean, would it be possible to close the lua subsystem after dt_control_shutdown() as we do for imageio and the rest of the gang?

in src/lua/init.c there are 2 functions, `dt_lua_finalize_early()` which sends an exit event to the scripts to let them know that darktable is exiting, then there is dt_lua_finalize()` which does the final closing (which happens to be right about `dt_lua_running()`.  I'd have to add some print statements and test it to see what happens.

**Comment by jenshannoschwalm** [comment]
Latest --force-push a) includes the pending jobs infrastructure in https://github.com/darktable-org/darktable/pull/18023/commits/5d6c6c9aee381a27243ade1ab04a9938cdf84994 .

Take **`Keep UI alive for a timeout before main window is closing`** as a *POC* for now, it's fully functional but as @ralfbrown mentioned it's by far not enough `obvious and unmissable`.

**Comment by ralfbrown** [comment]
My choice would be an exit screen and reparenting any active progress bars to the exit screen.  If an exit screen is not an option, restrict UI interactions to the progress bars (unhiding the left panel if necessary), preferably hiding or at least graying out all other panels and the center view to make it clear that the progress bars are the only UI elements still active.

Leavng the UI fully functional has an additional complication I didn't think to mention before - the user can start up additional background  jobs....

**Comment by jenshannoschwalm** [comment]
> Leavng the UI fully functional has an additional complication I didn't think to mention before - the user can start up additional background jobs....

Right, i also thought about that, we can and probably should use a `control->quitting` check in `dt_control_add_job()` to prohibit this. UI-wise the job-adding buttons might be greyed-out or whatever.



> My choice would be an exit screen and reparenting any active progress bars to the exit screen. If an exit screen is not an option, restrict UI interactions to the progress bars (unhiding the left panel if necessary), preferably hiding or at least graying out all other panels and the center view to make it clear that the progress bars are the only UI elements still active.

I so far just concentrated on the safety and correctness of pending jobs working in the background after the main window closing. I am absolutely not the UI-guy, i would love you or @dterrahe to take over on this part. Take the last commit as it is intended - a POC.

**Comment by jenshannoschwalm** [comment]
@zisoft just mentioning, i just found a way to trigger a mutex_destroy issue on release build - it now at least reports about such issues :-) Reproducer for now:
1. Export a number of images
2. Do that again on the same images
3. Discard the second backjob

**Comment by zisoft** [comment]
@jenshannoschwalm:
I just started my usual *rotate 10 images* test case. 

- select 10 images in lighttable
- click the *rotate 90 degrees* button

I get a reproducible crash as soon as dt starts to rotate the 3rd image:

Bildschirmfoto 2024-12-22 um 13 46 35

**Comment by jenshannoschwalm** [comment]
Ok there was a bad assert in dpthreads code, has been fixed there. For some reason the force-push of that commit doesn't show up in the above list so you really have to take the **full bunch** of 10 commits (or at least the first 9 :-) for testing.

Also added progress bar mutex fixing code in  https://github.com/darktable-org/darktable/pull/18023/commits/3be93ab4671130cf1c85f73f19ff1fb7ddbf513a

I started to test debug builds here systematically after your reports and can not find an issue while flipping, neither when doing them all, discarding them or processing in the dark.

As the release build also reports all pthread & mutex related errors now it might be worthwhile to also test that on your machine.

**Comment by zisoft** [comment]
MUCH better!

- started the rotate, let it finish, quit - OK
- started the rotate, quit after 2 images have been processed. The toast message shows the countdown, job finishes, darktable quits - OK
- started the rotate, quit after 2 images have been processed, cancel the job by pressing the X on the progress bar. darktable quits - OK

One issue:
Since there is no feedback when clicking the X I was not sure if that was registered and I clicked the X again. This gives a reproducible crash:

Bildschirmfoto 2024-12-22 um 18 25 52

I think there is another flag needed to prevent multiple cancellations.

**Comment by jenshannoschwalm** [comment]
> I think there is another flag needed to prevent multiple cancellations.

Exactly :-) I am already trying to find the "bad boy".

**Comment by jenshannoschwalm** [comment]
@zisoft @ralfbrown i got the bad boy for the mutex destroying issue and it really was a bad one. See latest `Fix mutex issues for cancel progress` commit including docs about why that was was done.

@wpferguson if you are going to test this and are looking out for lua issues, please note that the lua progress cancel callbacks don't destroy the progress - at least it looks to me like that ...

If you are going to test again please use the full fresh commits for this. This has been a very long bumpy ride and i certainly didn't want to have an *endless* list of commits :-)

**Comment by jenshannoschwalm** [comment]
There are still signal and clean-up issues in the darktable.lib system, it's safer to do that earlier so another round ...

**Comment by zisoft** [comment]
Next test results:

- rotate 10 images, let the job finish, quit - OK
- rotate 10 images, quit after job start, job finishes within the 30sec timeout - OK
- rotate 10 images, quit after job start, cancel jobs by pressing X - OK

This works fine now, no matter how often I click on the X - no more crashes.

But there must be something different between a forced cancellation and the timout cancellation. If I rotate 15 images and hit quit, the job does not finish within the 30sec timeout. This gives a crash on exit.

I noticed 3 different crashes in this scenario:

In the dbus routines (this is the most often)

Bildschirmfoto 2024-12-23 um 11 53 16

gui_module access during shutdown

Bildschirmfoto 2024-12-23 um 12 47 30

And rarely

Bildschirmfoto 2024-12-23 um 12 11 22


BTW: The toast is very hard to read with the transparent background:

Bildschirmfoto 2024-12-23 um 11 52 11

And maybe the text itself should be changed?
*you could discard those jobs within the next xx seconds, all remaining jobs will continue being processed after the main window has closed*
If I discard these jobs then... what remaining jobs?

**Comment by jenshannoschwalm** [comment]
Ok, I see the point about GUI stuff read from the pipe in different modules.

The time out closing is just a demo and test case for now how that works until someone does something really cool.

**Comment by jenshannoschwalm** [comment]
> But there must be something different between a forced cancellation and the timout cancellation. If I rotate 15 images and hit quit, the job does not finish within the 30sec timeout. This gives a crash on exit.

I don't understand this in detail. Could you elaborate what you are doing?

> In the dbus routines (this is the most often)

Don't compile unity so i'll have to test



> gui_module access during shutdown

Haven't seen that. Does this also happen without the latest commit with the delayed window closing?

**Comment by jenshannoschwalm** [comment]
> If I discard these jobs then... what remaining jobs?

There are none left, the window will close immediately after you clicked on the discard-x button.

**Comment by zisoft** [comment]
> > But there must be something different between a forced cancellation and the timout cancellation. If I rotate 15 images and hit quit, the job does not finish within the 30sec timeout. This gives a crash on exit.

> I don't understand this in detail. Could you elaborate what you are doing?

I rotate 15 images. The job starts and I quit. Now the timeout toast pops up. I can now cancel the running jobs by pressing the X on the progress bar. This successfully stops the jobs and darktable quits. That's OK.
Now the same test, rotate 15 images, quit. The timeout toast pops up. This time I just wait. The 30 seconds is not enough time to finish the jobs, so they get cancelled by the timeout (I think?). This time darktable crashes.

> > gui_module access during shutdown

> Haven't seen that. Does this also happen without the latest commit with the delayed window closing?

Will test and report.

> > If I discard these jobs then... what remaining jobs?

> There are none left, the window will close immediately after you clicked on the discard-x button.

I am referring to the toast text here. It says 
*you could discard those jobs within the next xx seconds, all remaining jobs will continue being processed after the main window has closed*

I don't understand that text. ...you can discard the jobs, all remaining jobs will continue being processed...
If I discard the jobs, what remaining jobs do you mean here? It is not understandable to the user.

File: src/common/dtpthread.h [hunk 2]
```diff
@@ -28,27 +28,36 @@
 #include <stdlib.h>
 #include <string.h>
 
-// #define MUTEX_REPORTING // uncomment to include runtime report of mutex problems also for release builds
+#define MUTEX_REPORTING // comment to exclude runtime reports of mutex problems from release builds
```
[the remark this answers is not in the document]
Comment (jenshannoschwalm) [reply src/common/dtpthread.h:31]: Ok. As a release version should be working ok so we don't need that. Something else for local build by people investigating problems.

**Comment by wpferguson** [comment]
Did some testing with Lua and exiting...

I have a long running script, responsive_cache, then generates a thumbtable and full screen cache on import or when invoked by shortcut.  It listens for the exit event, and runs a stop_job() callback to stop the processing.  I added print statements to see when the exit event was caught and when the job was stopped.

I tested by importing a filmroll (several hundred images), watching the job progress start, then hitting ctrl-q to exit darktable,  I added print statements to dt_lua_finalize_early() and dt_lua_finalize() to see when things were being done.

Remember in all this that Lua is single threaded, which does have an impact.

Test 1:

- Import a directory, let cache generation start, ctrl-q

Results

- GUI exited immediately
- script continued running and generated the complete cache set
- after the script completed the Lua exit processing took place.  As you can see from the log snipped below, the script held the thread so the finalize functions couldn't run.

```
    56.0724 LUA INFO: responsive_cache.lua:  processed 195 images in 19 seconds 
in dt_lua_finalize_early readying to fire the exit event
    56.0726 LUA INFO: callback: got exit signal 
exit event fired
waking up main context
in dt_lua_finalize getting ready to close lua state
closed luaA state
closed lua state
    57.7317 [pwstorage_new] Destroying context 0x602e7dd94f30
```

Lua supports cooperative multitasking where separate processes run under the control of a master process and voluntarily yield the thread.  darktable uses Lua in an event driven manner which doesn't permit this type of multitasking.  However, this development cycle I figured out how to do event driven cooperative multitasking, since I was tired of waiting for the cache to complete so I could do other things.  

Test 2

1. Import a directory of images with the cache running in multitasking mode.
2. watch the cache start
3. ctrl-q

Results

- GUI exited immediately
- responsize_cache yielded the thread 
- the event fired
- the event was caught
- the next responsive_cache yield Lua exited.

```
    51.3912 LUA DEBUG: callback: got message from responsive_cache to scheduler with message yield;131;170;3 
    51.3913 LUA DEBUG: callback: got message yield;131;170;3 from responsive_cache 
    51.3913 LUA INFO: scheduler.lua:  run queue length is 1 
    51.8921 LUA INFO: scheduler.lua:  run queue length is 0 
    51.8927 LUA DEBUG: scheduler.lua: remove_from_run_queue: 151: sending resume;131;170;3 to responsive_cache 
    51.8934 LUA INFO: responsive_cache.lua:  got message resume;131;170;3  resuming... 
in dt_lua_finalize_early readying to fire the exit event
    54.5155 LUA INFO: callback: got exit signal 
exit event fired
waking up main context
    54.5156 LUA DEBUG: callback: got message from responsive_cache to scheduler with message yield;171;210;3 
    54.5156 LUA DEBUG: callback: got message yield;171;210;3 from responsive_cache 
    54.5157 LUA INFO: scheduler.lua:  run queue length is 1 
in dt_lua_finalize getting ready to close lua state
closed luaA state
closed lua state

```

Buried in the documentation is a statement about a variable, `darktable.control.ending`, that gets set when darkable is starting to shut down.

Test 3

- Import a directory of images with the cache running in single threaded mode
- watch the cache start
- ctrl-q

Results

- GUI exited immediately
- caught darktable.control.ending and called the stop_job() function
- got several more darktable.control.ending messages, then an error about an invalid job.
- the lua finalize routines ran

```
    28.2678 LUA INFO: responsive_cache.lua:  creating thumbtable cache 
    31.6455 LUA dt.control.ending set, stopping job
    31.6455 LUA got exit
    31.6457 [dt_control_progress_destroy] progress at 0x7f3a180bf430
    31.7044 LUA dt.control.ending set, stopping job
    31.9062 LUA dt.control.ending set, stopping job
    32.1300 LUA dt.control.ending set, stopping job
    32.1928 LUA dt.control.ending set, stopping job
    32.2547 LUA dt.control.ending set, stopping job
    32.4533 LUA dt.control.ending set, stopping job
    33.3159 LUA ERROR : Accessing an invalid job
stack traceback:
	[C]: in ?
	[C]: in ?
	[C]: in metamethod 'newindex'
	...18023_thread_safety/lua/development/responsive_cache.lua:307: in upvalue 'generate_cache'
	...18023_thread_safety/lua/development/responsive_cache.lua:347: in field 'run_jobs'
	...18023_thread_safety/lua/development/responsive_cache.lua:381: in upvalue 'setup_jobs'
	...18023_thread_safety/lua/development/responsive_cache.lua:463: in function <...18023_thread_safety/lua/development/responsive_cache.lua:459>
	[C]: in ?
	[C]: in ?
in dt_lua_finalize_early readying to fire the exit event
    33.3160 LUA INFO: callback: got exit signal 
exit event fired
waking up main context
in dt_lua_finalize getting ready to close lua state
closed luaA state
closed lua state

```

Better coding would make the stop clean.


Let me know how you would like Lua shutdown to work, and I can add/enforce some coding standards, at least for the repository, to handle long running jobs.

**Comment by jenshannoschwalm** [comment]
@TurboGit i think i have sorted out the **safety requirements** and will propose another PR only including the technical parts for a safe shutdown. That will be much easier to check.

**Comment by jenshannoschwalm** [comment]
Closed this because of #18061 ... all still necessary from above will be done in single commit PR's after that has been merged hopefully.
